File: data/licenselynx-licenselynx/python/licenselynx/licenselynx.py
#
# SPDX-FileCopyrightText: Copyright 2025 Siemens AG
# SPDX-License-Identifier: BSD-3-Clause
#
import sys
from typing import Optional
from licenselynx.license_object import LicenseObject
from licenselynx.license_map_singleton import _LicenseMapSingleton
from licenselynx.quotes_handler import _QuotesHandler


class LicenseLynx:

    @staticmethod
    def map(license_name: str, risky: bool = False) -> Optional[LicenseObject]:
        """
        Maps license name to the canonical license identifier
        :param license_name: string of a license name
        :param risky: enable risky mappings
        :return: LicenseObject with the canonical license identifier and source, None if no license is found,
        or throws an exception if a runtime error occurs
        """
        try:
            license_name = _QuotesHandler().normalize_quotes(license_name)
            instance = _LicenseMapSingleton()

            license_object: Optional[LicenseObject] = instance.merged_data.stable_map.get(license_name)

            if not license_object and risky:
                license_object = instance.merged_data.risky_map.get(license_name)

            if not license_object:
                return None

            return license_object
        except Exception as e:
            raise e.with_traceback(sys.exc_info()[2])

File: data/licenselynx-licenselynx/go/licenselynx.go
// SPDX-FileCopyrightText: Copyright 2026 Siemens AG
// SPDX-License-Identifier: BSD-3-Clause

package licenselynx

//go:generate go run ./internal/codegen -input ../_support/merged_data.build.json -output .

type options struct {
	risky bool
	org   *Organization
}

// Option customizes Map behavior.
type Option func(*options)

// WithRisky enables lookup in the risky mappings.
func WithRisky() Option {
	return func(o *options) {
		o.risky = true
	}
}

// WithOrganization enables lookup in the organization-specific mappings.
func WithOrganization(org Organization) Option {
	return func(o *options) {
		o.org = &org
	}
}

// Map resolves a license name or identifier to its canonical representation.
// The lookup is performed in the following order:
// 1. Stable mappings (always)
// 2. Risky mappings (if WithRisky() is provided)
// 3. Organization-specific mappings (if WithOrganization(org) is provided)
//
// The function returns the resolved LicenseObject and a boolean indicating whether the mapping was successful.
// If the boolean is false, no suitable mapping was found and the returned LicenseObject will be empty.
func Map(licenseName string, opts ...Option) (LicenseObject, bool) {
	return mapWithLicenseMaps(licenseName, defaultLicenseMaps(), opts...)
}

// mapWithLicenseMaps is an internal helper that evaluates the provided options and performs the lookup in the provided license maps.
func mapWithLicenseMaps(licenseName string, maps licenseMaps, opts ...Option) (LicenseObject, bool) {
	resolvedOptions := options{}
	for _, opt := range opts {
		if opt != nil {
			opt(&resolvedOptions)
		}
	}

	normalizedLicenseName := normalizeQuotes(licenseName)

	if licenseObject, ok := maps.stable[normalizedLicenseName]; ok {
		return licenseObject, true
	}

	if resolvedOptions.risky {
		if licenseObject, ok := maps.risky[normalizedLicenseName]; ok {
			return licenseObject, true
		}
	}

	if resolvedOptions.org != nil {
		if licenseObject, ok := maps.organizationMap(*resolvedOptions.org)[normalizedLicenseName]; ok {
			return licenseObject, true
		}
	}

	return LicenseObject{}, false
}
